osgar/drivers/canserial.py

```diff
@@ -79,15 +79,102 @@ def split_buffer(data):
                 return data[2+size:], data[:2+size]
         return data, b''  # no complete packet available yet
 
+    #################### TODO refactor ###################
+    def send_data(self, module_id, data):
+        self.bus.publish('raw', CAN_packet(module_id, data))
+
+    def read_packet(self):
+        while True:
+            self.buf, packet = self.split_buffer(self.buf)
+            if len(packet) > 0:
+                msg_id = ((packet[0]) << 3) | (((packet[1]) >> 5) & 0x1f)
+                return msg_id, packet[2:]
+            else:
+                dt, channel, data = self.bus.listen()
+                if channel == 'raw':
+                    self.buf += data
+                else:
+                    print('Ignoring', channel)
+
+    def reset_modules(self):
+        """Reset all modules"""
+        self.send_data(0, [129,0]) # reset all
+
+        ackBootup = [] # Heart Beat 0, bootup, after reset
+        ackPreop = [] # HB 127
+        ackOp = [] # HB 5
+
+        while len(ackBootup) == 0 or len(ackBootup) > len(ackPreop):
+            msg_id, data = self.read_packet()
+            if (msg_id & 0xF80) == 0x700:
+                nodeID = msg_id & 0x7F
```

Comment:
Maybe having less code like this would make it more readable. I am not very sure why 0x700 is important. I guess 0xf80 and 0x7f serve to parse subset of bits from the `msg_id` but I have no idea why. I see that 0x7f is actually 0b1111111 and 0xf80 is 0b111110000000 which for me make it more obvious.

Reply:
0x700 is actually 0b1110 (`"{:b}".format(0x700>>7)`). So... the code
```python
(msg_id & 0xF80) == 0x700
```
is actually 
```python
(msg_id >> 8) == 0b111
```
??

I am still not sure why that is important though.

Reply:
I think we need more than the link (please don't paste the table into the code). Can we have some comments noting the intention of the code? Something like `if packet comes from...??` etc. It will make the code easier to review and contribute to. 

I still don't know what `if (msg_id & 0xF80) == 0x700:` does. Even if I study the CANOpen page (which should not be needed for the code review), I see that msg_id is supposed to be 11 bits but 0xF80 has 12 bits. Shouldn't the mask to get the first 4 bits from the original 11 bits be 0x780?

I guess you are trying to check if the message is a heart beat message? Did I guess correctly? Isn't the way to do that just `if (msg_id & 0x700):`? Putting comment `is heartbeat message` after the if would be the best in this case (if I guessed correctly).

I guess what I am saying is, that I don't understand the code and I feel there is not enough information for me to get that understanding. Without it, I am not able to help.

Reply:
Unfortunately I have to agree with you ;). There is a first step towards better understanding = describing what numbers/masks means. The `restart_modules()` is still not nice - a unittest would be nice ... TODO.

Reply:
Much better, thanks.